### PR TITLE
Many loss functions 

### PR DESCRIPTION
**Pull request by smolix**
Added a plethora of loss functions (probably more to come) to make losses a winning thing. In particular, added many classical scalar loss functions, more max-margin losses and some losses for the dual Ali-Slivey estimation.

### CODE REVIEW

File: python/mxnet/gluon/loss.py [hunk 1]
```diff
@@ -385,3 +377,511 @@ def hybrid_forward(self, F, data, label,
...
+    """
+    def __init__(self, rho=1, weight=None, batch_axis=0, **kwargs):
+        super(Huber, self).__init__(weight, batch_axis, **kwargs)
+        self._rho = rho
+
+    def hybrid_forward(self, F, output, label, sample_weight=None):
+        label = _reshape_label_as_output(F, output, label)
+        loss = F.abs(output - label)
+        loss = ((loss > self._rho) * (loss - 0.5 * self._rho) +
+               (0.5/self._rho) * (loss <= self._rho) * loss**2)
```
Comment (szha) [inline python/mxnet/gluon/loss.py]: F.where(loss > self._rho, (loss - 0.5 * self._rho), (0.5/self._rho)*loss**2) so that mask is calculated once.

Comment (smolix) [reply python/mxnet/gluon/loss.py]: Elegant! Now I learned that we have a where operator ...
